libkbfs/md_journal.go

```diff
@@ -0,0 +1,507 @@
+		dir:    dir,
+		j:      makeMdIDJournal(codec, journalDir),
+	}
+	return journal
+}
+
+// The functions below are for building various paths.
+
+func (j mdJournal) mdsPath() string {
+	return filepath.Join(j.dir, "mds")
+}
+
+func (j mdJournal) mdPath(id MdID) string {
+	idStr := id.String()
+	return filepath.Join(j.mdsPath(), idStr[:4], idStr[4:])
+}
+
+// getMD verifies the MD data and the writer signature (but not the
+// key) for the given ID and returns it.
+func (j mdJournal) getMD(id MdID) (*BareRootMetadata, error) {
+	// Read file.
+
+	path := j.mdPath(id)
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	var rmd BareRootMetadata
+	err = j.codec.Decode(data, &rmd)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check integrity.
+
+	mdID, err := j.crypto.MakeMdID(&rmd)
+	if err != nil {
+		return nil, err
+	}
+
+	if id != mdID {
+		return nil, fmt.Errorf(
+			"Metadata ID mismatch: expected %s, got %s", id, mdID)
+	}
+
+	err = rmd.VerifyWriterMetadata(j.codec, j.crypto)
+	if err != nil {
+		return nil, err
+	}
+
+	return &rmd, nil
+}
+
+// putMD stores the given metadata under its ID, if it's not already
+// stored.
+func (j mdJournal) putMD(
+	rmd *BareRootMetadata, currentUID keybase1.UID,
+	currentVerifyingKey VerifyingKey) (MdID, error) {
+	err := rmd.IsValidAndSigned(
+		j.codec, j.crypto, currentUID, currentVerifyingKey)
+	if err != nil {
+		return MdID{}, err
+	}
+
+	id, err := j.crypto.MakeMdID(rmd)
+	if err != nil {
+		return MdID{}, err
+	}
+
+	_, err = j.getMD(id)
+	if os.IsNotExist(err) {
+		// Continue on.
+	} else if err != nil {
+		return MdID{}, err
+	} else {
+		// Entry exists, so nothing else to do.
+		return MdID{}, nil
+	}
+
+	path := j.mdPath(id)
+
+	err = os.MkdirAll(filepath.Dir(path), 0700)
+	if err != nil {
+		return MdID{}, err
+	}
+
+	buf, err := j.codec.Encode(rmd)
+	if err != nil {
+		return MdID{}, err
+	}
+
+	err = ioutil.WriteFile(path, buf, 0600)
+	if err != nil {
+		return MdID{}, err
+	}
+
+	return id, nil
+}
+
+func (j mdJournal) getHeadHelper() (ImmutableBareRootMetadata, error) {
+	headID, err := j.j.getLatest()
+	if err != nil {
+		return ImmutableBareRootMetadata{}, err
+	}
+	if headID == (MdID{}) {
+		return ImmutableBareRootMetadata{}, nil
+	}
+	head, err := j.getMD(headID)
+	if err != nil {
+		return ImmutableBareRootMetadata{}, err
+	}
+	return MakeImmutableBareRootMetadata(head, headID), nil
+}
+
+func (j mdJournal) checkGetParams(currentUID keybase1.UID) (
+	ImmutableBareRootMetadata, error) {
+	head, err := j.getHeadHelper()
+	if err != nil {
+		return ImmutableBareRootMetadata{}, err
+	}
+
+	if head != (ImmutableBareRootMetadata{}) {
+		ok, err := isReader(currentUID, head.BareRootMetadata)
+		if err != nil {
+			return ImmutableBareRootMetadata{}, err
+		}
+		if !ok {
+			// TODO: Use a non-server error.
+			return ImmutableBareRootMetadata{},
+				MDServerErrorUnauthorized{}
+		}
+	}
+
+	return head, nil
+}
+
+func (j mdJournal) convertToBranch(
```

Comment:
It seems possible for a `put()` call to come in halfway through this call, and get written to disk as a merged revision before all of its predecessors are marked as unmerged.  Its `PrevRoot` would also be wrong at that point. I guess when the flush finally gets to that wrong MD, it would fail to put on the server due to server-side checking, but I'm not sure if it would fail due to a revision conflict, or if it could fail due to `PrevRoot` checking or something like that.

Have you thought about that case?  Should we have some locking here so that branch conversion happens more atomically?